File: app/schemas/perfis.py
"""Schemas dos perfis de adaptacao: neurodivergencia, Big Five e
preferencias de acessibilidade."""

import uuid
from datetime import datetime

from pydantic import BaseModel, Field, field_validator


class CondicaoPublica(BaseModel):
    id: uuid.UUID
    codigo: str
    nome: str
    descricao: str | None

    model_config = {"from_attributes": True}


class RegistrarPerfilAlunoRequest(BaseModel):
    condicoes_codigos: list[str] = Field(
        default_factory=list,
        description="Codigos de app/GET /condicoes-neurodivergencia. Lista vazia = nenhuma "
        "condicao identificada nesta versao.",
    )
    observacoes: str | None = Field(default=None, max_length=2000)
    aceite_consentimento: bool = Field(
        ...,
        description=(
            "Consentimento explicito e especifico para tratamento deste dado sensivel de "
            "saude, distinto do consentimento geral de cadastro."
        ),
    )


class PerfilAlunoResponse(BaseModel):
    id: uuid.UUID
    aluno_id: uuid.UUID
    versao: int
    observacoes: str | None
    criado_por_id: uuid.UUID
    criado_em: datetime
    condicoes: list[CondicaoPublica]


class QuestaoTIPI(BaseModel):
    ordem: int
    texto: str


class BigFiveRespostasRequest(BaseModel):
    respostas: list[int] = Field(
        ..., min_length=10, max_length=10, description="10 respostas, escala 1 (discordo "
        "totalmente) a 7 (concordo totalmente), na ordem de GET /big-five/questionario."
    )

    @field_validator("respostas")
    @classmethod
    def _validar_escala(cls, respostas: list[int]) -> list[int]:
        if any(r < 1 or r > 7 for r in respostas):
            raise ValueError("Cada resposta deve estar entre 1 e 7.")
        return respostas


class BigFiveScores(BaseModel):
    abertura: float
    conscienciosidade: float
    extroversao: float
    amabilidade: float
    neuroticismo: float


class PerfilBigFiveResponse(BaseModel):
    id: uuid.UUID
    aluno_id: uuid.UUID
    versao: int
    criado_em: datetime
    scores: BigFiveScores
    instrumento: str


class PreferenciasAcessibilidadeRequest(BaseModel):
    fonte_legivel: bool = False
    alto_contraste: bool = False
    tempo_extra_percentual: int = Field(default=0, ge=0, le=200)
    leitura_voz_alta: bool = False
    reducao_estimulos: bool = False
    tamanho_fonte: str = Field(default="medio", pattern="^(pequeno|medio|grande|extra_grande)$")


class PreferenciasAcessibilidadeResponse(PreferenciasAcessibilidadeRequest):
    usuario_id: uuid.UUID

    model_config = {"from_attributes": True}

File: frontend/src/lib/api/perfis.ts
import { apiClient } from './client'
import type {
  BigFiveRespostasRequest,
  CondicaoPublica,
  PerfilAlunoResponse,
  PerfilBigFiveResponse,
  PreferenciasAcessibilidadeRequest,
  PreferenciasAcessibilidadeResponse,
  QuestaoTIPI,
  RegistrarPerfilAlunoRequest,
} from './types'

export async function listarCondicoes(): Promise<CondicaoPublica[]> {
  const { data } = await apiClient.get<CondicaoPublica[]>('/condicoes-neurodivergencia')
  return data
}

export async function obterQuestionarioBigFive(): Promise<QuestaoTIPI[]> {
  const { data } = await apiClient.get<QuestaoTIPI[]>('/big-five/questionario')
  return data
}

export async function enviarBigFive(payload: BigFiveRespostasRequest): Promise<PerfilBigFiveResponse> {
  const { data } = await apiClient.post<PerfilBigFiveResponse>('/me/big-five', payload)
  return data
}

export async function obterPreferenciasAcessibilidade(): Promise<PreferenciasAcessibilidadeResponse> {
  const { data } = await apiClient.get<PreferenciasAcessibilidadeResponse>(
    '/me/preferencias-acessibilidade',
  )
  return data
}

export async function atualizarPreferenciasAcessibilidade(
  payload: PreferenciasAcessibilidadeRequest,
): Promise<PreferenciasAcessibilidadeResponse> {
  const { data } = await apiClient.put<PreferenciasAcessibilidadeResponse>(
    '/me/preferencias-acessibilidade',
    payload,
  )
  return data
}

export async function registrarPerfilAluno(
  alunoId: string,
  payload: RegistrarPerfilAlunoRequest,
): Promise<PerfilAlunoResponse> {
  const { data } = await apiClient.post<PerfilAlunoResponse>(`/alunos/${alunoId}/perfil`, payload)
  return data
}

export async function obterPerfilAluno(alunoId: string): Promise<PerfilAlunoResponse> {
  const { data } = await apiClient.get<PerfilAlunoResponse>(`/alunos/${alunoId}/perfil`)
  return data
}

export async function obterHistoricoPerfilAluno(alunoId: string): Promise<PerfilAlunoResponse[]> {
  const { data } = await apiClient.get<PerfilAlunoResponse[]>(`/alunos/${alunoId}/perfil/historico`)
  return data
}

export async function obterBigFiveDeAluno(alunoId: string): Promise<PerfilBigFiveResponse> {
  const { data } = await apiClient.get<PerfilBigFiveResponse>(`/alunos/${alunoId}/big-five`)
  return data
}
